### ISSUE TITLE
C++: use C++17, integrate clang-tidy, clang-format

### ISSUE THREAD

**Issue by dirkolbrich** [IfcOpenShell/IfcOpenShell#3686]
In connection of an issue regarding the [update of cmake related settings and structure](https://github.com/IfcOpenShell/IfcOpenShell/issues/3670) I would like to propose a general update of the C++ code base and related tooling starting with branch v0.8.0 (-> repo/branch structure, another topic IMV).

- Update to use C++17
- Integrate clang-tidy tooling
- Integrate clang-format

## C++17

In my view an obvious update. Most dependency packages are already on or support C++17.

## clang-tidy

Provide static linting in the tool chain to make the overall project less error prone.

Currently on branch v0.7.0 or v0.8.0 if I activate clang-tidy the warnings light up like a Christmas tree. So it probably would be best to discuss which settings would be the most appropriate. I know for my self that I don’t like [modernize-use-trailing-return-type](https://clang.llvm.org/extra/clang-tidy/checks/modernize/use-trailing-return-type.html) for readability reasons.

So maybe it would be best to start with some basic checks and then slowly integrate single `modernise-` checks and work up IfcOS module by module. 

```clang-tidy
Checks: „cert-*, bugprone-*’
```

## clang-format

Enforce a unified code style across the project to aid in readability.

I know this is a very optionated topic. But coming from Golang I wish C++ had something like a default style.
I played a bit with the [Clang-Format Detector](https://clangpowertools.com/) (open source tool for Windows to detect clang-format styles setting from existing code bases and good playground for alternative setting), but haven’t found a good starting point yet. What are your preferences here?

Referencing: https://github.com/IfcOpenShell/IfcOpenShell/issues/21, https://github.com/IfcOpenShell/IfcOpenShell/issues/2691

**Comment by aothms** [comment]
I fully agree. I haven't experimented with clang-tidy yet, but running this in CI/CD doesn't hurt if we see conformance as a moving target. I also wanted to use a formatting utility after the v08 rewrite was done and then strictly conform to it. Actually I don't care too much about the formatting conventions we choose, but in my mind I had the following ideas:

**`snake_case` variable/function/class naming everywhere**

 to (a) conform to C++ STL / boost and (most of) python naming conventions (b) easily disambiguate with Ifc type names.

I know not everybody is a fan of this, I don't care too much if we choose differently.


**braces**

Over time I've found that readability sometimes conflicts with consistency. For example:

Generally I prefer opening braces on same line:

      void do_something_else() {
      }
      ...
      if (something) {
        do_something_else();
      }

but with constructors you'd typically write on a new line
    
    class foo {
      int a_, b_;
      foo(int a, int b)
        : a_(a)
        , b_(b)
      {}

or multiline if statements

    if ((some_long_variable == some_other_long_variable) &&
        (yet_another_long_variable != some_other_long_variable))
    {

So probably opening brace on a new line is the sane thing to do. Even if I find it eats too many lines.

**Pointers directly adjacent to type**

Visually I prefer 

    int* a;

over

    int *a;

But you'd have to really be careful, because syntactically the `*` binds to the right.

    int* a, b;

Means

    int *a;
    int b;

---

Anyway, this is all to say I don't think there is right and wrong in these kind of discussions. As long as we choose something and stick to it consistently we win.

**Comment by Moult** [comment]
> As long as we choose something and stick to it consistently we win.

This :)

**Comment by dirkolbrich** [comment]
Thanks. some good starting points where I think we agree.

**Agreed**: opening braces on same line:
```cpp
if (something) { ... }
```

For the multiline `if` statement with long conditionals I would rewrite it to use locally scoped boolean variables to make the if-statement shorter and thus better readable:
```cpp
bool variables_equal = (some_long_variable == some_other_long_variable);
bool other_variables_not_equal = (yet_another_long_variable != some_other_long_variable);
if (variables_equal && other_variables_not_equal) { ... }
```

**Agreed**: pointers directly adjacent to type, as the pointer is part of the type, not the variable:
```cpp
int* a;
```
Here I follow [Bjarnes explanation](https://www.stroustrup.com/bs_faq2.html#whitespace).

Declaring multiple variables in a single statement [should be avoided anyways](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Res-name-one) (and should be [initialised on declaration!](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#es20-always-initialize-an-object)), so it would be better to write:
```cpp
int* a = nullptr;
int* b = nullptr;
```

So, how do we start? Would you at some point rebase v0.8.0 on v0.7.0? And after that I would add some tidy and format in a separate branch based on v0.8.0 to test the best approach.
Starting right away on v0.7.0 doesn't make much sense to me if there are already a lot of merge conflicts between v0.8 and v0.7. I don't want to add too much noise to that.

**Comment by aothms** [comment]
> For the multiline if statement with long conditionals I would rewrite it to use locally scoped boolean variables to make the if-statement shorter and thus better readable:

In quite a few cases we depend on the short circuiting behaviour of boolean conjunction. Separating them out to loose variables would be undefined behaviour or throws exceptions.

```C++
// contrived example
std::optional<int> a;
if (a && *a == 1) {

}
```

Actual example:

https://github.com/IfcOpenShell/IfcOpenShell/blob/0a00e529c3fe1b4f2002ae2227ef3ac212a40a6a/src/ifcgeom_schema_agnostic/wire_utils.cpp#L457

You can't call `ecc.Distance(1)` when `ecc.NbExtrema() == 0`.

In principle I agree though.

> Here I follow [Bjarnes explanation](https://www.stroustrup.com/bs_faq2.html#whitespace).

ok won't argue with Stroustrup

> So, how do we start? Would you at some point rebase v0.8.0 on v0.7.0?

v0.7.0 would be merged in v0.8.0. I have already done that some months ago fighting through the most painful conflicts. But things have of course again diverged. The main issue is that v0.8.0 is still not fully feature complete wrt to the more obscure things (that nobody will ever likely notice) and likely still performs slightly worse in the default scenario. I want to prioritise this, but I keep being flooded with other work (the ifc4.3 iso submission deadline this week for ex).

If you want you can start on v0.8 with the parts `ifcgeom/kernels/cgal` (since this does not exist in v0.7 so can't have conflicts) and `ifcparse/` (I don't think there has been any work on this for about a year on v0.7 so also there nothing can conflict).

**Comment by dirkolbrich** [comment]
Sounds like a plan. 👍🏻 

I'd like to start with `/ifcparse` as a nice separate unit (and I don't know much about cgal). Yet there are some changes v0.7.0 -> v0.8.0. Running `git diff v0.7.0 v0.8.0 --compact-summary -- src/ifcparse` gives this:

```shell
src/ifcparse/Argument.h              |    23 +-
src/ifcparse/Ifc2x3-schema.cpp       | 21988 ++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++-------------------------------------------------------------------------------------------------------------
src/ifcparse/Ifc2x3.cpp              |  2472 +++++---------------------
src/ifcparse/Ifc2x3.h                |   224 +--
src/ifcparse/Ifc4-schema.cpp         | 26089 ++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++------------------------------------------------------------------------------------------------------------------------------------
src/ifcparse/Ifc4.cpp                |  3054 ++++++--------------------------
src/ifcparse/Ifc4.h                  |   240 +--
src/ifcparse/Ifc4x1-schema.cpp       | 26770 ++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++---------------------------------------------------------------------------------------------------------------------------------------
src/ifcparse/Ifc4x1.cpp              |  3085 ++++++--------------------------
src/ifcparse/Ifc4x1.h                |   240 +--
src/ifcparse/Ifc4x2-schema.cpp       | 27400 +++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++------------------------------------------------------------------------------------------------------------------------------------------
src/ifcparse/Ifc4x2.cpp              |  3226 ++++++----------------------------
src/ifcparse/Ifc4x2.h                |   240 +--
src/ifcparse/Ifc4x3-definitions.h    |    50 -
src/ifcparse/Ifc4x3.cpp              |   176 +-
src/ifcparse/Ifc4x3.h                |   237 +--
src/ifcparse/Ifc4x3_add1-schema.cpp  | 29818 +++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++-----------------------------------------------------------------------------------------------------------------------------------------------------
src/ifcparse/Ifc4x3_add1.cpp         |  3747 ++++++---------------------------------
src/ifcparse/Ifc4x3_add1.h           |   237 +--
src/ifcparse/Ifc4x3_rc1-schema.cpp   | 29874 +++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++------------------------------------------------------------------------------------------------------------------------------------------------------
src/ifcparse/Ifc4x3_rc1.cpp          |  3730 ++++++---------------------------------
src/ifcparse/Ifc4x3_rc1.h            |   254 +--
src/ifcparse/Ifc4x3_rc2-schema.cpp   | 30053 ++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++-------------------------------------------------------------------------------------------------------------------------------------------------------
src/ifcparse/Ifc4x3_rc2.cpp          |  3768 +++++++--------------------------------
src/ifcparse/Ifc4x3_rc2.h            |   255 +--
src/ifcparse/Ifc4x3_rc3-schema.cpp   | 29871 +++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++------------------------------------------------------------------------------------------------------------------------------------------------------
src/ifcparse/Ifc4x3_rc3.cpp          |  3774 ++++++---------------------------------
src/ifcparse/Ifc4x3_rc3.h            |   240 +--
src/ifcparse/Ifc4x3_rc4-schema.cpp   | 29926 +++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++------------------------------------------------------------------------------------------------------------------------------------------------------
src/ifcparse/Ifc4x3_rc4.cpp          |  3793 ++++++----------------------------------
src/ifcparse/Ifc4x3_rc4.h            |   240 +--
src/ifcparse/Ifc4x3_tc1-schema.cpp   | 29719 ++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++-----------------------------------------------------------------------------------------------------------------------------------------------------
src/ifcparse/Ifc4x3_tc1.cpp          |  3747 ++++++---------------------------------
src/ifcparse/Ifc4x3_tc1.h            |   237 +--
src/ifcparse/IfcBaseClass.h          |    25 +
src/ifcparse/IfcFile.h               |    13 +-
src/ifcparse/IfcHierarchyHelper.cpp  |    20 +-
src/ifcparse/IfcParse.cpp            |     2 +-
src/ifcparse/IfcParse.h              |     1 +
src/ifcparse/IfcSchema.h             |    17 +
src/ifcparse/aggregate_of_instance.h |     9 +-
src/ifcparse/macros.h                |    10 +-
src/ifcparse/parse_ifcxml.cpp        |     2 +-
43 files changed, 151431 insertions(+), 167465 deletions(-)
```
So, would it be possible to merge just a single subdirectory from v.0.7 to v0.8?

**Comment by dirkolbrich** [comment]
https://github.com/IfcOpenShell/IfcOpenShell/pull/3751

**Comment by aothms** [comment]
You're right. I attempted to do so. With:

1. find commits in 0.7 that affect src/ifcparse

    git log $(git merge-base origin/v0.7.0 origin/v0.8.0)..origin/v0.7.0 --reverse --pretty="%h" src/ifcparse

2. apply them with cherry-pick

    git cherry-pick 885bca086 63eaa88d3 765214418 30d9743ae b33e58f40 aab6a21bf f5c102402 ed5d28de5 97c6b99d5 0cdf72fde 997ebe48d 0bb9bcd9f 65cb47e2d 2d330f4c9 6816ad5c0 7541293a3

3. Out of these only `65cb47e2d` affected both ifcparse/ and ifcgeom/ so this conflicted the cherry-pick I skipped it and then ran

    git show 65cb47e2d aggregate_of_instance.h | git apply

(I'm writing this down mostly for myself in case I encounter this mess later again)

There are indeed some changes remaining now, but they look benign / intended to me.

**Comment by aothms** [comment]
This probably means though that ifcgeom/ is broken now, because we had this as an important backwards incompatible change https://github.com/IfcOpenShell/IfcOpenShell/issues/2805

**Comment by dirkolbrich** [comment]
Thanks for the update. I think I can now safely start on v0.8 without too many v0.7 conflicts. With the latests commit [43b525e](https://github.com/IfcOpenShell/IfcOpenShell/commit/43b525e7e7004b49048938c38169d84be3ae7e17) the changes looks like this:

```shell
 src/ifcparse/Argument.h             |    23 +-
 src/ifcparse/Ifc2x3-schema.cpp      | 21988 ++++++++++++++++++++++++++++++++++-------------------------------
 src/ifcparse/Ifc2x3.cpp             |  2294 +------
 src/ifcparse/Ifc4-schema.cpp        | 26089 +++++++++++++++++++++++++++++++++++++++--------------------------------------
 src/ifcparse/Ifc4.cpp               |  2874 ++-------
 src/ifcparse/Ifc4x1-schema.cpp      | 26770 ++++++++++++++++++++++++++++++++++++++++---------------------------------------
 src/ifcparse/Ifc4x1.cpp             |  2905 ++-------
 src/ifcparse/Ifc4x2-schema.cpp      | 27400 +++++++++++++++++++++++++++++++++++++++++----------------------------------------
 src/ifcparse/Ifc4x2.cpp             |  3046 ++-------
 src/ifcparse/Ifc4x3_add1-schema.cpp | 29818 +++++++++++++++++++++++++++++++++++++++++++++-------------------------------------------
 src/ifcparse/Ifc4x3_add1.cpp        |  3571 ++---------
 src/ifcparse/Ifc4x3_rc1-schema.cpp  | 29874 +++++++++++++++++++++++++++++++++++++++++++++-------------------------------------------
 src/ifcparse/Ifc4x3_rc1.cpp         |  3542 ++---------
 src/ifcparse/Ifc4x3_rc2-schema.cpp  | 30053 ++++++++++++++++++++++++++++++++++++++++++++++-------------------------------------------
 src/ifcparse/Ifc4x3_rc2.cpp         |  3580 ++---------
 src/ifcparse/Ifc4x3_rc3-schema.cpp  | 29871 +++++++++++++++++++++++++++++++++++++++++++++-------------------------------------------
 src/ifcparse/Ifc4x3_rc3.cpp         |  3598 ++---------
 src/ifcparse/Ifc4x3_rc4-schema.cpp  | 29926 +++++++++++++++++++++++++++++++++++++++++++++-------------------------------------------
 src/ifcparse/Ifc4x3_rc4.cpp         |  3617 ++---------
 src/ifcparse/Ifc4x3_tc1-schema.cpp  | 29719 +++++++++++++++++++++++++++++++++++++++++++++-------------------------------------------
 src/ifcparse/Ifc4x3_tc1.cpp         |  3571 ++---------
 src/ifcparse/IfcBaseClass.h         |    25 +
 src/ifcparse/IfcParse.h             |     1 +
 src/ifcparse/IfcSchema.h            |    17 +
 src/ifcparse/macros.h               |    10 +-
 25 files changed, 149432 insertions(+), 164750 deletions(-)
```

where only ` src/ifcparse/Argument.h` and `src/ifcparse/macros.h` seem to have some ambiguous changes:

```git diff v0.7.0..v0.8.0 -- src/ifcparse/Argument.h```:
```diff
diff --git a/src/ifcparse/Argument.h b/src/ifcparse/Argument.h
index b4935c58b..e67a5b9bd 100644
--- a/src/ifcparse/Argument.h
+++ b/src/ifcparse/Argument.h
@@ -27,32 +27,15 @@
 #include <algorithm>
 
 #include "ifc_parse_api.h"
-
-#include "../ifcparse/aggregate_of_instance.h"
+#include "ArgumentType.h"
 
 #include <boost/shared_ptr.hpp>
 #include <boost/dynamic_bitset.hpp>
 #include <boost/logic/tribool.hpp>
 
-/*
-namespace boost {
-       namespace logic {
-               class tribool {
-
-               };
-       }
-}
-*/
-
-class Argument;
-
 class aggregate_of_instance;
 class aggregate_of_aggregate_of_instance;
 
-class IfcEntityInstanceData;
-namespace IfcParse {
-       class IfcFile;
-}
 namespace IfcUtil {
        class IfcBaseClass;
 
@@ -76,11 +59,11 @@ public:
        virtual operator std::vector<double>() const;
        virtual operator std::vector<std::string>() const;
        virtual operator std::vector<boost::dynamic_bitset<> >() const;
-       virtual operator aggregate_of_instance::ptr() const;
+       virtual operator boost::shared_ptr<aggregate_of_instance>() const;
 
        virtual operator std::vector< std::vector<int> >() const;
        virtual operator std::vector< std::vector<double> >() const;
-       virtual operator aggregate_of_aggregate_of_instance::ptr() const;
+       virtual operator boost::shared_ptr<aggregate_of_aggregate_of_instance>() const;
 
        virtual bool isNull() const = 0;
        virtual unsigned int size() const = 0;
```

```git diff v0.7.0..v0.8.0 -- src/ifcparse/macros.h```:
```diff
diff --git a/src/ifcparse/macros.h b/src/ifcparse/macros.h
index 8db482967..cbf801243 100644
--- a/src/ifcparse/macros.h
+++ b/src/ifcparse/macros.h
@@ -20,15 +20,17 @@
 #ifndef IFCOPENSHELL_MACROS_H
 #define IFCOPENSHELL_MACROS_H
 
-#define MAKE_TYPE_NAME__(a, b) a ## b
-#define MAKE_TYPE_NAME_(a, b) MAKE_TYPE_NAME__(a, b)
-#define MAKE_TYPE_NAME(t) MAKE_TYPE_NAME_(t, IfcSchema)
+#define POSTFIX_SCHEMA__(a, b) a ## _ ## b
+#define POSTFIX_SCHEMA_(a, b) POSTFIX_SCHEMA__(a, b)
+#define POSTFIX_SCHEMA(t) POSTFIX_SCHEMA_(t, IfcSchema)
 
 #define STRINGIFY_(x) #x
 #define STRINGIFY(x) STRINGIFY_(x)
 
-#define MAKE_INIT_FN__(a, b) init_ ## a ## b
+#define MAKE_INIT_FN__(a, b) init_ ## a ## _ ## b
 #define MAKE_INIT_FN_(a, b) MAKE_INIT_FN__(a, b)
 #define MAKE_INIT_FN(t) MAKE_INIT_FN_(t, IfcSchema)
 
+
+
 #endif
\ No newline at end of file
```

**Comment by dirkolbrich** [comment]
closed with https://github.com/IfcOpenShell/IfcOpenShell/pull/3751

**Comment by aothms** [comment]
Yes these changes look ok

**Comment by dirkolbrich** [comment]
> This probably means though that ifcgeom/ is broken now, because we had this as an important backwards incompatible change #2805

Yep, on first compile I get some errors for not matching constructors:
```shell
/Users/dirkolbrich/Code/IfcOS/IfcOpenShell/src/ifcgeom/Serialization/schema/Serialization.cpp:404:10: error: no matching constructor for initialization of 'Ifc2x3::IfcTrimmedCurve'
        c = new IfcSchema::IfcTrimmedCurve(base, trim1, trim2, true, IfcSchema::IfcTrimmingPreference::IfcTrimmingPreference_PARAMETER);
                ^                          ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
<command line>:20:19: note: expanded from macro 'IfcSchema'
#define IfcSchema Ifc2x3
                  ^
/Users/dirkolbrich/Code/IfcOS/IfcOpenShell/src/ifcgeom/Serialization/schema/../../../ifcparse/Ifc2x3.h:29196:5: note: candidate constructor not viable: no known conversion from 'shared_ptr<aggregate_of_instance>' to 'shared_ptr<aggregate_of<IfcTrimmingSelect>>' for 2nd argument
    IfcTrimmedCurve (::Ifc2x3::IfcCurve* v1_BasisCurve, aggregate_of< ::Ifc2x3::IfcTrimmingSelect >::ptr v2_Trim1, aggregate_of< ::Ifc2x3::IfcTrimmingSelect >::ptr v3_Trim2, bool v4_SenseAgreement, ::Ifc2x3::IfcTrimmingPreference::Value v5_MasterRepresentation);
    ^
/Users/dirkolbrich/Code/IfcOS/IfcOpenShell/src/ifcgeom/Serialization/schema/../../../ifcparse/Ifc2x3.h:29195:5: note: candidate constructor not viable: requires single argument 'e', but 5 arguments were provided
    IfcTrimmedCurve (IfcEntityInstanceData* e);
    ^
/Users/dirkolbrich/Code/IfcOS/IfcOpenShell/src/ifcgeom/Serialization/schema/../../../ifcparse/Ifc2x3.h:29176:21: note: candidate constructor (the implicit copy constructor) not viable: requires 1 argument, but 5 were provided
class IFC_PARSE_API IfcTrimmedCurve : public  IfcBoundedCurve {
                    ^
/Users/dirkolbrich/Code/IfcOS/IfcOpenShell/src/ifcgeom/Serialization/schema/../../../ifcparse/Ifc2x3.h:29176:21: note: candidate constructor (the implicit move constructor) not viable: requires 1 argument, but 5 were provided
/Users/dirkolbrich/Code/IfcOS/IfcOpenShell/src/ifcgeom/Serialization/schema/Serialization.cpp:645:20: error: no matching constructor for initialization of 'Ifc2x3::IfcShellBasedSurfaceModel'
                        items->push(new IfcSchema::IfcShellBasedSurfaceModel(shells->generalize()));
                                        ^                                    ~~~~~~~~~~~~~~~~~~~~
<command line>:20:19: note: expanded from macro 'IfcSchema'
#define IfcSchema Ifc2x3
                  ^
/Users/dirkolbrich/Code/IfcOS/IfcOpenShell/src/ifcgeom/Serialization/schema/../../../ifcparse/Ifc2x3.h:15773:21: note: candidate constructor (the implicit copy constructor) not viable: no known conversion from 'aggregate_of_instance::ptr' (aka 'shared_ptr<aggregate_of_instance>') to 'const IfcShellBasedSurfaceModel' for 1st argument
class IFC_PARSE_API IfcShellBasedSurfaceModel : public  IfcGeometricRepresentationItem {
                    ^
/Users/dirkolbrich/Code/IfcOS/IfcOpenShell/src/ifcgeom/Serialization/schema/../../../ifcparse/Ifc2x3.h:15773:21: note: candidate constructor (the implicit move constructor) not viable: no known conversion from 'aggregate_of_instance::ptr' (aka 'shared_ptr<aggregate_of_instance>') to 'IfcShellBasedSurfaceModel' for 1st argument
/Users/dirkolbrich/Code/IfcOS/IfcOpenShell/src/ifcgeom/Serialization/schema/../../../ifcparse/Ifc2x3.h:15779:5: note: candidate constructor not viable: no known conversion from 'aggregate_of_instance::ptr' (aka 'shared_ptr<aggregate_of_instance>') to 'IfcEntityInstanceData *' for 1st argument
    IfcShellBasedSurfaceModel (IfcEntityInstanceData* e);
    ^
/Users/dirkolbrich/Code/IfcOS/IfcOpenShell/src/ifcgeom/Serialization/schema/../../../ifcparse/Ifc2x3.h:15780:5: note: candidate constructor not viable: no known conversion from 'shared_ptr<aggregate_of_instance>' to 'shared_ptr<aggregate_of<IfcShell>>' for 1st argument
    IfcShellBasedSurfaceModel (aggregate_of< ::Ifc2x3::IfcShell >::ptr v1_SbsmBoundary);
    ^
/Users/dirkolbrich/Code/IfcOS/IfcOpenShell/src/ifcgeom/Serialization/schema/Serialization.cpp:678:52: error: no matching constructor for initialization of 'Ifc2x3::IfcGeometricCurveSet'
                                        IfcSchema::IfcGeometricCurveSet* curves = new IfcSchema::IfcGeometricCurveSet(edges);
                                                                                      ^                               ~~~~~
<command line>:20:19: note: expanded from macro 'IfcSchema'
#define IfcSchema Ifc2x3
                  ^
/Users/dirkolbrich/Code/IfcOS/IfcOpenShell/src/ifcgeom/Serialization/schema/../../../ifcparse/Ifc2x3.h:18835:21: note: candidate constructor (the implicit copy constructor) not viable: no known conversion from 'aggregate_of_instance::ptr' (aka 'shared_ptr<aggregate_of_instance>') to 'const IfcGeometricCurveSet' for 1st argument
class IFC_PARSE_API IfcGeometricCurveSet : public  IfcGeometricSet {
                    ^
/Users/dirkolbrich/Code/IfcOS/IfcOpenShell/src/ifcgeom/Serialization/schema/../../../ifcparse/Ifc2x3.h:18835:21: note: candidate constructor (the implicit move constructor) not viable: no known conversion from 'aggregate_of_instance::ptr' (aka 'shared_ptr<aggregate_of_instance>') to 'IfcGeometricCurveSet' for 1st argument
/Users/dirkolbrich/Code/IfcOS/IfcOpenShell/src/ifcgeom/Serialization/schema/../../../ifcparse/Ifc2x3.h:18839:5: note: candidate constructor not viable: no known conversion from 'aggregate_of_instance::ptr' (aka 'shared_ptr<aggregate_of_instance>') to 'IfcEntityInstanceData *' for 1st argument
    IfcGeometricCurveSet (IfcEntityInstanceData* e);
    ^
/Users/dirkolbrich/Code/IfcOS/IfcOpenShell/src/ifcgeom/Serialization/schema/../../../ifcparse/Ifc2x3.h:18840:5: note: candidate constructor not viable: no known conversion from 'shared_ptr<aggregate_of_instance>' to 'shared_ptr<aggregate_of<IfcGeometricSetSelect>>' for 1st argument
    IfcGeometricCurveSet (aggregate_of< ::Ifc2x3::IfcGeometricSetSelect >::ptr v1_Elements);
    ^
```

**Comment by aothms** [comment]
Should be fixed with 5bead4b219dcfc71f10ab8d3d267c4b559fbbfd7
